server/src/main/java/org/elasticsearch/index/mapper/JsonFieldParser.java

```diff
@@ -31,53 +31,117 @@
             parser::getTokenLocation);
 
+        ContentPath path = new ContentPath();
         List<IndexableField> fields = new ArrayList<>();
-        int openObjects = 1;
 
+        parseObject(parser, path, fields);
+        return fields;
+    }
+
+    private void parseObject(XContentParser parser,
+                             ContentPath path,
+                             List<IndexableField> fields) throws IOException {
+        String currentName = null;
         while (true) {
-            if (openObjects == 0) {
-                return fields;
+            XContentParser.Token token = parser.nextToken();
+            if (token == XContentParser.Token.END_OBJECT) {
+                return;
             }
 
+            if (token == XContentParser.Token.FIELD_NAME) {
+                currentName = parser.currentName();
+            } else if (token == XContentParser.Token.START_OBJECT) {
+                assert currentName != null;
+                path.add(currentName);
+                parseObject(parser, path, fields);
+                path.remove();
+            } else if (token == XContentParser.Token.START_ARRAY) {
+                assert currentName != null;
+                parseArray(parser, path, currentName, fields);
+            }  else if (token.isValue()) {
+                String value = parser.text();
+                addField(path, currentName, value, fields);
+            } else if (token == XContentParser.Token.VALUE_NULL) {
+                String value = fieldType.nullValueAsString();
+                if (value != null) {
+                    addField(path, currentName, value, fields);
+                }
+            }
+        }
+    }
+
+    private void parseArray(XContentParser parser,
+                             ContentPath path,
+                             String currentName,
+                             List<IndexableField> fields) throws IOException {
+        while (true) {
             XContentParser.Token token = parser.nextToken();
-            assert token != null;
+            if (token == XContentParser.Token.END_ARRAY) {
+                return;
+            }
 
             if (token == XContentParser.Token.START_OBJECT) {
-                openObjects++;
-            } else if (token == XContentParser.Token.END_OBJECT) {
-                openObjects--;
+                path.add(currentName);
+                parseObject(parser, path, fields);
+                path.remove();
             } else if (token.isValue()) {
                 String value = parser.text();
-                addField(value, fields);
+                addField(path, currentName, value, fields);
             } else if (token == XContentParser.Token.VALUE_NULL) {
                 String value = fieldType.nullValueAsString();
                 if (value != null) {
-                    addField(value, fields);
+                    addField(path, currentName, value, fields);
                 }
             }
```

Comment:
Oops, I think this is actually a bug, since an array of arrays is valid JSON and should be accepted. Will fix.

I agree it's a good idea to add an `else` with an exception, so we'll fail fast when encountering something unexpected rather than attempt to proceed in a potentially wrong state.